Refactor TVDB client and local cache

JoeBiellik opened the pull request:
This is my work so far on completely refactoring the handing of TheTVDB data. This isn't ready to merge and is in initial stages, but I would very much like any testing or feedback.

I've implemented my changes in a separate project as to not break the main application and for better separation of the UI and logic layers. My plan is to refactor and move all the code from the main project until only the UI is left. I've changed targeting to .NET 4.6 as it has the same system requirements as 4.5 but adds some nice features.

So far I've implemented a simple TVDB API client and the underlying serialization logic to save the cache to a file. The major change here is that I'm using direct serialization rather than parsing (#259).
DTO classes almost exactly represent TheTVDB results and the API client specifies the API version to use, so I don't foresee changing schemas to be an issue, and rather than storing all the returned data, the objects are converted to more convenient formats and only keep whats used.

Currently, the application passes around a lot of references to the cache instance, I've changed this to be a singleton which can be used anywhere. All the cached shows in the singleton should be fully thread safe and parallelism is achieved with native datatypes avoiding any locking. I'm also heavily using the [TPL](https://docs.microsoft.com/en-us/dotnet/standard/parallel-programming/task-parallel-library-tpl) so there should be zero blocking of the calling thread which will avoid all the current UI lockups and allow instant cancels.

From my initial tests, this is a lot faster than the current implementation. My ``TheTVDB.xml`` is 85MB and has over 60 million lines, the JSON equivalent, with all the same data is 35MB and 40M lines. It's very hard to compare the cache load time as the current code is very interconnected with the UI, but with 500 shows the JSON takes about a quarter of a second to deserialize on my desktop. I've done some memory profiling but again its hard because of the UI.

Error handling, progress reporting and languages aren't fully implemented and I wanted to get feedback before I wrote any unit tests.

For quick testing you can drop this code at the top of the ``Main()`` entrypoint of the application:
```cs
// Random sample shows
var ids = new[] { 104581, 110381, 104641, 108611, 112141, 112871, 114701, 118421, 121361, 123411, 123581, 124051, 125441, 128711, 129261, 130111, 134241, 135291, 137591, 139941, 140681, 141521, 142561, 147071, 148581, 150051, 152831, 153021, 155201, 158661, 159431, 159591 };

// Run async as sync
Task.Run(async () =>
{
    // Trigering this should instantly cancel operations
    CancellationTokenSource cts = new CancellationTokenSource();

    // Deserialize to singleton, create a fresh instance if it doesn't exist
    await TVDB.Load("tvdb.json");

    // Do multiple TheTVDB API requests at once in parallel
    TVDB.Instance.Threads = 10;

    // Add some sample shows, download all metatdata
    await ids.ForEachAsync(async id => await TVDB.Instance.Add(id, cts.Token), TVDB.Instance.Threads); // Add 10 at a time

    // "Force refresh" entire cache
    //await TVDB.Instance.Refresh(cts.Token);

    // Process TheTVDB update feed and download as needed
    await TVDB.Instance.Update(cts.Token);

    // Serialize and save the cache instance
    await TVDB.Save("tvdb.json");
}).GetAwaiter().GetResult();
```

I would very much appreciate any thoughts on this code or how we handle the metadata in general.

MarkSummerville commented:
Great - I'll take a look over the next few days. My initial thoughts are that it 'feels' a lot nicer. Some of the async stuff if a little beyond my skill level right now to be able to provide any useful feedback, but I'm hoping that it cleans up whole tranches of existing code.  When it is more complete I can unpick the current code to identify business rules that may not be obvious that we can  double check are included in the new version.

JoeBiellik commented:
I've added an example form project to give you an idea of how you interact with the core asynchronously. It's a very basic recreation of the main form but should demonstrate non-blocking UI updates.

The add show form is an example of task cancellation. Rather than using a webpage I've implemented a standard control to display the show. I'm not happy with the abstraction between the cache of shows in memory and the UI bindings, so I may look at using native binding to controls with a ``BindingList`` and implementing ``IPropertyChange ``.

I've also completely rebuilt ``ApplicationBase`` as to unbind startup operations from the UI thread and because we were previously using ``WindowsFormsApplicationBase`` which has a huge overhead the way it was being used.

JoeBiellik commented:
I've ported more of the UI and begun implementing refactored ``DownloadIdentifers`` including an improved template engine as per #301.

I'll merge this in soon so I have a base to work from, even if it never directly makes it to master.
